narwhal/primary/src/state_handler.rs

```diff
@@ -267,11 +530,14 @@ impl StateHandler {
                         SystemMessage::DkgConfirmation(conf) => {
                             randomness_state.add_confirmation(conf.clone())
                         }
+                        SystemMessage::RandomnessSignature(round, _sig) => {
+                            randomness_state.update_randomness_round(*round + 1).await;
+                        }
                     }
                 }
-                // Advance the random beacon protocol if possible after each certificate.
+                // Advance the random beacon DKG protocol if possible after each certificate.
                 // TODO: Implement/audit crash recovery for random beacon.
-                randomness_state.advance(&self.tx_system_messages).await;
+                randomness_state.advance_dkg().await;
```

Comment:
For now if a node crashes it will not participate in random beacon until next epoch. I plan to come back and impl this before launch, presumably will have to either store some state on disk or else replay the SystemMessage history from epoch start on startup